btp-document/Smart-Contracts-ICON.md

```diff
@@ -0,0 +1,387 @@
+
+```shell
+source rpc.sh
+rpcch icon
+rpcks $GOLOOP_KEY_STORE $GOLOOP_KEY_SECRET
+goloop rpc sendtx deploy ./javascore/parachain-BMV-optimized.jar \
+    --content_type application/java --step_limit 13610920001 \
+    --param relayMtaOffset=${Relaychain Offset} \
+    --param paraMtaOffset=${Parachain Offset} \
+    --param bmc=$(cat bmc.$(rpcch)) \
+    --param net=$(cat net.btp.icon) \
+    --param mtaRootSize=${MTA root size} \
+    --param mtaCacheSize=${MTA caches size} \
+    --param mtaIsAllowNewerWitness=${Allow MTA newer witness} \
+    --param relayLastBlockHash=${Relaychain last blockhash} \
+    --param paraLastBlockHash=${Parachain last blockhash} \
+    --param encodedValidators=${Base64 of encodedValidators} \
+    --param relayEventDecoderAddress=$(cat kusamaDecoder.icon) \
+    --param paraEventDecoderAddress=$(cat moonriverDecoder.icon) \
+    --param relayCurrentSetId=${setID} \
+    --param paraChainId=${Parachain ID}
+     | jq -r . > tx.bmv.$(rpcch)
+```
+
+- Extract address of ***BMV*** contract after deployment as:
+
+```shell
+goloop rpc txresult $(cat tx.bmv.$(rpcch)) | jq -r .scoreAddress > bmv.$(rpcch)
+```
+
+### 4. Deploy IRC31Token Contract
+
+____
+
+- Run these commands to deploy ***IRC31Token*** on ICON Network. Again, these commands should be run in a goloop container.
+
+```shell
+source rpc.sh
+rpcch icon
+rpcks $GOLOOP_KEY_STORE $GOLOOP_KEY_SECRET
+goloop rpc sendtx deploy ./javascore/irc31-0.1.0-debug.jar \
+    --content_type application/java --step_limit 13610920001 | jq -r . > tx.irc31token.$(rpcch)
+```
+
+- Extract address of ***IRC31Token*** contract after deployment as:
+
+```shell
+goloop rpc txresult $(cat tx.irc31token.$(rpcch)) | jq -r .scoreAddress > irc31token.$(rpcch)
+```
+
+### 5. Deploy NativeCoinBSH Contract
+
+____
+
+- Run these commands to deploy ***NativeCoinBSH*** on ICON Network. Again, these commands should be run in a goloop container.
+
+```shell
+source rpc.sh
+rpcch icon
+rpcks $GOLOOP_KEY_STORE $GOLOOP_KEY_SECRET
+goloop rpc sendtx deploy ./javascore/nativecoin-0.1.0-debug.jar \
+    --content_type application/java --step_limit 13610920001 \
+    --param _bmc=$(cat bmc.$(rpcch)) \
+    --param _irc31=$(cat irc31token.$(rpcch)) \
+    --param _name=ICX | jq -r . > tx.nativeCoinBsh.$(rpcch)
+```
+
+- Extract address of ***NativeCoinBSH*** contract after deployment as:
+
+```shell
+goloop rpc txresult $(cat tx.nativeCoinBsh.$(rpcch)) | jq -r .scoreAddress > nativeCoinBsh.$(rpcch)
+```
+
+### 6. Deploy FeeAggregation Contract
+
+____
+
+- Run below commands to compile Java files and generate requiring `.jar` files:
+
+```shell
+$ cd $CONFIG_DIR/btp/javascore/fee_aggregation
+
+$ ./gradlew build
+
+$ ./gradlew optimizedJar
+```
+
+- Finally, deploy ***FeeAggregation*** contract on ICON Network as follow:
+
+```shell
+./gradlew deployToLocal -PkeystoreName=<your_wallet_json> -PkeystorePass=<password>
```

Comment:
This command does not go through for me. I get the following error:

```joriss-MacBook-Pro:fee_aggregation joris$ ./gradlew deployToLocal -PkeystoreName=../../../keystore.json -PkeystorePass=116d1bb85afe692d

> Task :deployToLocal FAILED
>>> deploy to http://localhost:9080/api/v3/icon
>>> optimizedJar = /Users/joris/Documents/work/btp_doc/INode/btp/javascore/fee_aggregation/build/libs/fee-aggregation-system-1.0-optimized.jar
>>> keystore = ../../../keystore.json

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':deployToLocal'.
> foundation.icon.icx.transport.jsonrpc.RpcError: E0006:InvalidParameter(type=Address,json="hx00000000000000000000000000000000000000000")

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 954ms
1 actionable task: 1 executed
```

Reply:
The code base was updated to adapt `gochain`. I noticed this error yesterday and already told them to update new instructions. Stay tuned